Initial implementation (#1)
yousty/pg_eventstore: docs/events_and_streams.md

# The description of Event and Stream definitions

> `pg_eventstore` provides classes to manipulate the data in the database. The most important are:

Reviewer: Misleading, you do not manipulate, you just append. So maybe better:

`pg_eventstore` provides classes to prepare the events to be inserted into the eventstore. The most important are:

Author: Yeah, thanks.

- `PgEventstore::Event` class which represents an event object
- `PgEventstore::Stream` class which represents a stream object

## Event object and its defaults

`PgEventstore::Event` has the following attributes:

- `id` - String(UUIDv4, optional). If no provided - the value will be autogenerated.
- `type` - String(optional). Default is an event's class name.

> - `global_position` - Integer(optional, read only). Event's position in "all" stream. Manually assigning this attribute will affect on nothing. It is internally set when reading events from the database.

Reviewer: ```suggestion
- `global_position` - Integer(optional, read only). Event's global position in the eventstore (the "all" stream position in the popular EventstoreDB). Manually assigning this attribute has no effect. It is internally set when reading events from the database.
```

> - `global_position` - Integer(optional, read only). Event's global position in the eventstore (the "all" stream position in the popular EventstoreDB). Manually assigning this attribute has no effect. It is internally set when reading events from the database.

Reviewer: ```suggestion
- `global_position` - Integer(optional, read only). Event's global position in the eventstore, aka the "all" stream position (inspired by the popular EventstoreDB). Manually assigning this attribute has no effect. It is internally set when reading events from the database.
```

> - `stream` - PgEventstore::Stream(optional, read only). A Stream an event belongs to, see description bellow. Manually assigning this attribute will affect on nothing. It is internally set when appending an event to the given stream or when reading events from the database.

Reviewer: ```suggestion
- `stream` - PgEventstore::Stream(optional, read only). A Stream an event belongs to, see description below. Manually assigning this attribute has no effect. It is internally set when appending an event to the given stream or when reading events from the database.
```

- `stream_revision` - Integer(optional, read only). A revision of an event inside its stream.

> - `data` - Hash(optional). Event's data. Usually you may want to put here the data which is directly related to an event's type. For example, if you have `DescriptionChanged` event class, then you may want to have a description value in the `data` attribute. Example: `DescriptionChanged.new(data: { 'description' => 'Description of something', 'post_id' => 1 })`

Reviewer: ```suggestion
- `data` - Hash(optional). Event's payload data. For example, if you have a `DescriptionChanged` event class, then you may want to have a description value in the event payload data. Example: `DescriptionChanged.new(data: { 'description' => 'Description of something', 'post_id' => SecureRandom.uuid })`
```

> - `metadata` - Hash(optional). Event's metadata. Usually you may want to have some "system" information about an event. Simply saying - and info which doesn't meaningfully suitable to be placed under `data` attribute.

Reviewer: ```suggestion
- `metadata` - Hash(optional). Event metadata. Event meta information which is not part of an events data payload. Example: `{ published_by: publishing_user.id }`
```

> - `link_id` - String(UUIDv4, optional, read only). If an event is a link event(link events are simply pointers to another events) - this attribute will be containing an `id` of an original event. Manually assigning this attribute will affect on nothing. It is internally set when appending an event to the given stream or when reading events from the database.

Reviewer: ```suggestion
- `link_id` - String(UUIDv4, optional, read only). If an event is a link event (link events are pointers to other events), this attribute contains the `id` of the original event. Manually assigning this attribute has no effect. It is internally set when appending an event to the given stream or when reading events from the database.
```

> - `created_at` - Time(optional, read only). Database's timestamp when an event was appended to a stream. You may want to put your own timestamp into a `metadata` attribute - it may be useful when migrating between different databases. Manually assigning this attribute will affect on nothing. It is internally set when appending an event to the given stream or when reading events from the database.

Reviewer: ```suggestion
- `created_at` - Time(optional, read only). Database's timestamp when an event was appended to a stream. You may want to put your own timestamp into a `metadata` attribute - it may be useful when migrating between different databases. Manually assigning this attribute has no effect. It is internally set when appending an event to the given stream or when reading events from the database.
```

Example:

```ruby
PgEventstore::Event.new(data: { 'foo' => 'bar' })
```

## Stream object

> To be able to manipulate a stream - you have to compute a stream's object first. It can be achieved by using `PgEventstore::Stream` class. Here is a description of its attributes:

Reviewer: ```suggestion
To be able to manipulate a stream, you have to compute a stream's object first. It can be achieved by using the `PgEventstore::Stream` class. Here is a description of its attributes:
```

- `context` - String(required). A Bounded Context, read more [here](https://martinfowler.com/bliki/BoundedContext.html). Values which start from `$` sign are reserved by `pg_eventstore`. Such contexts can't be used to append events.
- `stream_name` - String(required). A stream name
- `stream_id` - String(required). A stream id

> - `id` - Integer(optional, read only). Internal id. It is set when a stream is returned from database as a part of deserialization process. Manually assigning this attribute will affect on nothing.

Reviewer: ```suggestion
- `id` - Integer(optional, read only). Internal id. It is set when a stream is returned from the database as part of the deserialization process. Manually assigning this attribute has no effect.
```

> - `stream_revision` - Integer(optional, read only). Current stream's revision. You can rely on this value when setting `:expected_revision` option when appending events to a stream. It is set when a stream is returned from database as a part of deserialization process. Manually assigning this attribute will affect on nothing.

Reviewer: ```suggestion
- `stream_revision` - Integer(optional, read only). Current stream's revision. You can rely on this value when setting the `:expected_revision` option when appending events to a stream. It is set when a stream is returned from the database a part of the deserialization process. Manually assigning this attribute has no effect.
```

Example:

```ruby
PgEventstore::Stream.new(context: 'Sales', stream_name: 'Customer', stream_id: '1')
```

> There is a special case, called "all" stream. You can get this object by calling `PgEventstore::Stream.all_stream` method. Read more about "all" stream in `Reading from the "all" stream` section of [Reading events](reading_events.md) chapter.

Reviewer: ```suggestion
There is a special stream, called the "all" stream. You can get this object by calling the`PgEventstore::Stream.all_stream` method. Read more about the "all" stream in the `Reading from the "all" stream` section of [Reading events](reading_events.md) chapter.
```
